File: layoutlib-api/src/main/java/com/android/ide/common/rendering/api/ResourceReference.java
/*
 * Copyright (C) 2011 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.android.ide.common.rendering.api;

import com.android.annotations.NonNull;
import com.android.annotations.concurrency.Immutable;
import com.android.resources.ResourceType;
import com.android.resources.ResourceUrl;
import com.google.common.base.MoreObjects;
import java.io.Serializable;

/**
 * A resource reference, contains the namespace, type and name. Can be used to look for resources in
 * a resource repository.
 *
 * <p>This is an immutable class.
 */
@Immutable
public class ResourceReference implements Serializable {
    @NonNull private final ResourceType mResourceType;
    @NonNull private final ResourceNamespace mNamespace;
    @NonNull private final String mName;

    public ResourceReference(
            @NonNull ResourceNamespace namespace,
            @NonNull ResourceType resourceType,
            @NonNull String name) {
        this.mNamespace = namespace;
        this.mResourceType = resourceType;
        this.mName = name;
    }

    /**
     * Used by layoutlib still. TODO(namespaces): remove this.
     *
     * @deprecated
     */
    @Deprecated
    public ResourceReference(@NonNull String name, boolean isFramework) {
        this(ResourceNamespace.fromBoolean(isFramework), ResourceType.LAYOUT, name);
    }

    @Deprecated
    public ResourceReference(
            @NonNull ResourceType type, @NonNull String name, boolean isFramework) {
        this(ResourceNamespace.fromBoolean(isFramework), type, name);
    }

    /** Returns the name of the resource, as defined in the XML. */
    @NonNull
    public String getName() {
        return mName;
    }

    @NonNull
    public ResourceType getResourceType() {
        return mResourceType;
    }

    @NonNull
    public ResourceNamespace getNamespace() {
        return mNamespace;
    }

    /**
     * Returns whether the resource is a framework resource (<code>true</code>) or a project
     * resource (<code>false</code>).
     *
     * @deprecated all namespaces should be handled not just "android:".
     */
    @Deprecated
    public final boolean isFramework() {
        return mNamespace == ResourceNamespace.ANDROID;
    }

    @NonNull
    public ResourceUrl getResourceUrl() {
        return ResourceUrl.create(mNamespace.getPackageName(), mResourceType, mName);
    }

    /**
     * Returns a {@link ResourceUrl} that can be used to refer to this resource from the given
     * namespace. This means the namespace part of the {@link ResourceUrl} will be null if the
     * context namespace is the same as the namespace of this resource.
     *
     * <p>This method assumes no namespace prefixes (aliases) are defined, so the returned {@link
     * ResourceUrl} will use the full package name of the target namespace, if necessary. Most use
     * cases should attempt to call the overloaded method instead and provide a {@link
     * ResourceNamespace.Resolver} from the XML element where the {@link ResourceUrl} will be used.
     *
     * @see #getRelativeResourceUrl(ResourceNamespace, ResourceNamespace.Resolver)
     */
    @NonNull
    public ResourceUrl getRelativeResourceUrl(@NonNull ResourceNamespace context) {
        return getRelativeResourceUrl(context, ResourceNamespace.Resolver.EMPTY_RESOLVER);
    }

    /**
     * Returns a {@link ResourceUrl} that can be used to refer to this resource from the given
     * namespace. This means the namespace part of the {@link ResourceUrl} will be null if the
     * context namespace is the same as the namespace of this resource.
     *
     * <p>This method uses the provided {@link ResourceNamespace.Resolver} to find the short prefix
     * that can be used to refer to the target namespace. If it is not found, the full package name
     * is used.
     */
    @NonNull
    public ResourceUrl getRelativeResourceUrl(
            @NonNull ResourceNamespace context, @NonNull ResourceNamespace.Resolver resolver) {
        String namespaceString;
        if (mNamespace.equals(context)) {
            namespaceString = null;
        } else {
            String prefix = resolver.uriToPrefix(mNamespace.getXmlNamespaceUri());
            if (prefix != null) {
                namespaceString = prefix;
            } else {
                namespaceString = mNamespace.getPackageName();
            }
        }

        return ResourceUrl.create(namespaceString, mResourceType, mName);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        ResourceReference reference = (ResourceReference) o;

        if (mResourceType != reference.mResourceType) return false;
        if (!mNamespace.equals(reference.mNamespace)) return false;
        if (!mName.equals(reference.mName)) return false;

        return true;
    }

    @Override
    public int hashCode() {
        int result = mResourceType.hashCode();
        result = 31 * result + mNamespace.hashCode();
        result = 31 * result + mName.hashCode();
        return result;
    }

    @Override
    public String toString() {
        return MoreObjects.toStringHelper(this)
                .add("namespace", mNamespace)
                .add("type", mResourceType)
                .add("name", mName)
                .toString();
    }
}

File: lint/libs/lint-api/src/main/java/com/android/tools/lint/client/api/ResourceReference.kt
/*
 * Copyright (C) 2016 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.android.tools.lint.client.api

import com.android.SdkConstants
import com.android.SdkConstants.ANDROID_PKG
import com.android.SdkConstants.ID_PREFIX
import com.android.SdkConstants.NEW_ID_PREFIX
import com.android.resources.ResourceType
import com.android.tools.lint.detector.api.isKotlin
import com.android.tools.lint.detector.api.stripIdPrefix
import com.google.common.base.Joiner
import com.intellij.psi.PsiArrayType
import com.intellij.psi.PsiClass
import com.intellij.psi.PsiField
import com.intellij.psi.PsiJavaFile
import com.intellij.psi.PsiModifier
import com.intellij.psi.PsiType
import com.intellij.psi.PsiVariable
import com.intellij.psi.util.PsiTreeUtil
import org.jetbrains.uast.UCallExpression
import org.jetbrains.uast.UElement
import org.jetbrains.uast.UExpression
import org.jetbrains.uast.UQualifiedReferenceExpression
import org.jetbrains.uast.UResolvable
import org.jetbrains.uast.USimpleNameReferenceExpression
import org.jetbrains.uast.UVariable
import org.jetbrains.uast.asQualifiedPath
import org.jetbrains.uast.getContainingUFile
import org.jetbrains.uast.getQualifiedParentOrThis
import org.jetbrains.uast.java.JavaUDeclarationsExpression

/**
 * A reference to an Android resource in the AST; the reference may not be qualified.
 * For example, in the below, the `foo` reference on the right hand side of
 * the assignment can be resolved as an [ResourceReference].
 * <pre>
 * import my.pkg.R.string.foo;
 * ...
 * int id = foo;
</pre> *
 */
class ResourceReference(
    val node: UExpression,
    // getPackage() can be empty if not a package-qualified import (e.g. android.R.id.name).
    val `package`: String,
    val type: ResourceType,
    val name: String,
    val heuristic: Boolean = false
) {

    internal val isFramework: Boolean
        get() = `package` == ANDROID_PKG

    companion object {
        private fun toAndroidReference(expression: UQualifiedReferenceExpression): ResourceReference? {
            val path = expression.asQualifiedPath() ?: return null

            var packageNameFromResolved: String? = null

            val containingClass =
                PsiTreeUtil.getParentOfType(expression.resolve(), PsiClass::class.java)
            if (containingClass != null) {
                val containingClassFqName = containingClass.qualifiedName
                if (containingClassFqName != null) {
                    val i = containingClassFqName.lastIndexOf(".R.")
                    if (i >= 0) {
                        packageNameFromResolved = containingClassFqName.substring(0, i)
                    }
                }
            }

            val size = path.size
            if (size < 3) {
                return null
            }

            val r = path[size - 3]
            if (r != SdkConstants.R_CLASS) {
                return null
            }

            val packageName = if (packageNameFromResolved != null)
                packageNameFromResolved
            else Joiner.on('.').join(path.subList(0, size - 3))

            val type = path[size - 2]
            val name = path[size - 1]

            val resourceType = ResourceType.fromClassName(type) ?: return null
            return ResourceReference(expression, packageName, resourceType, name)
        }

        @JvmStatic
        fun get(element: UElement): ResourceReference? {
            // Optimization for Java: instead of resolving the field just peek at the reference
            // and pick out the resource type and name from the context.
            // This also lets us pick up resource references even when the R fields don't
            // resolve (e.g. when there are symbol or source error problems.)
            if (element is UQualifiedReferenceExpression) {
                val ref = toAndroidReference(element)
                if (ref != null) {
                    return ref
                }
            }

            val declaration = when (element) {
                is UVariable -> element.psi
                is UResolvable -> (element as UResolvable).resolve()
                else -> return null
            }

            if (declaration == null && element is USimpleNameReferenceExpression) {
                // R class can't be resolved in tests so we need to use heuristics to calc the reference
                val maybeQualified = (element as UExpression).getQualifiedParentOrThis()
                if (maybeQualified is UQualifiedReferenceExpression) {
                    val ref = toAndroidReference(maybeQualified)
                    if (ref != null) {
                        return ref
                    }
                }
            }

            if (declaration !is PsiVariable) {
                // Synthetic import?
                // In the IDE, this will resolved into XML PSI. Attempt to use reflection to
                // pick out the relevant attribute.
                if (declaration != null &&
                    declaration::class.java.name == "com.intellij.psi.impl.source.xml.XmlAttributeValueImpl" &&
                    element is UExpression
                ) {
                    try {
                        val method = declaration::class.java.getDeclaredMethod("getValue")
                        val value = method.invoke(declaration)?.toString() ?: ""
                        if (value.startsWith(ID_PREFIX) || value.startsWith(NEW_ID_PREFIX)) {
                            return ResourceReference(
                                element,
                                "",
                                ResourceType.ID,
                                stripIdPrefix(value)
                            )
                        }
                    } catch (ignore: Throwable) {
                    }
                }

                val parent = element.uastParent
                if (parent is UQualifiedReferenceExpression && parent.selector === element) {
                    // synthetic import reference is usually not qualified
                    return null
                }
                if (parent is UCallExpression && parent.classReference === element) {
                    return null
                }

                if (declaration == null &&
                    // In the IDE we have proper reference resolving for synthetic imports
                    !LintClient.isStudio &&
                    element is USimpleNameReferenceExpression &&
                    isKotlin(element.sourcePsi) &&
                    element.identifier != "it"
                ) {
                    // If we have any synthetic imports in this class, this unresolved symbol is
                    // probably referring to it
                    element.getContainingUFile()?.imports?.forEach {
                        val expression = it.importReference as? USimpleNameReferenceExpression
                        val resolved = expression?.resolvedName
                        if (resolved != null &&
                            (
                                resolved.startsWith("import kotlinx.android.synthetic.") ||
                                    resolved.startsWith("kotlinx.android.synthetic.")
                                )
                        ) {
                            return ResourceReference(
                                element,
                                "",
                                ResourceType.ID,
                                element.identifier,
                                heuristic = true
                            )
                        }
                    }
                }

                return null
            }

            val variable = declaration as PsiVariable?
            if (variable !is PsiField ||
                (variable.type != PsiType.INT && !isIntArray(variable.type)) ||
                // Note that we don't check for PsiModifier.FINAL; in library projects
                // the R class fields are deliberately not made final such that their
                // values can be substituted when all the resources are merged together
                // in the app module and unique id's can be assigned for all resources
                !variable.hasModifierProperty(PsiModifier.STATIC)
            ) {
                return null
            }

            val resTypeClass = variable.containingClass
            if (resTypeClass == null || !resTypeClass.hasModifierProperty(PsiModifier.STATIC)) {
                return null
            }

            val rClass = resTypeClass.containingClass
            if (rClass == null || rClass.containingClass != null) {
                return null
            } else {
                val className = rClass.name
                if (!("R" == className || "R2" == className)) { // R2: butterknife library
                    return null
                }
            }

            val packageName = (rClass.containingFile as PsiJavaFile).packageName
            if (packageName.isEmpty()) {
                return null
            }

            val resourceType =
                ResourceType.fromClassName(resTypeClass.name ?: return null) ?: return null
            val resourceName = variable.name
            val node: UExpression = when (element) {
                is UExpression -> element
                is UVariable -> JavaUDeclarationsExpression(null, listOf(element))
                else -> throw IllegalArgumentException("element must be an expression or a UVariable")
            }

            return ResourceReference(node, packageName, resourceType, resourceName)
        }

        /**
         * Returns true if the type represents an int array (int[]), which is the
         * type of styleable R fields.
         */
        private fun isIntArray(type: PsiType): Boolean =
            (type as? PsiArrayType)?.componentType == PsiType.INT
    }
}
